### PR TITLE
metrics: add structured format for Hubble metrics and options.

### PR DESCRIPTION
**Pull request by rectified95**
_Re-submitting PR #32294 now that the requested Hubble metrics server test was added in https://github.com/cilium/cilium/pull/34325_
_All the work was authored by @vakalapa. In the commit I added, I addressed the small comments left on it and simplified the test code by removing some indirection when passing in `ContextOptionConfig`._

This change adds a structured internal format for metrics and metrics options, instead of the existing string representation.

First PR in the series to implement:
https://github.com/cilium/design-cfps/blob/main/hubble/CFP-30788-managing-hubble-metrics-cardinality.md 

CFP: https://github.com/cilium/design-cfps/pull/29

Will follow up with code to derive this config from a new ConfigMap.

Checklist:
- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

### CODE REVIEW

**Comment by rectified95** [comment]
FYI @chancez

File: pkg/hubble/metrics/api/api.go [hunk 1]
```diff
@@ -146,3 +128,101 @@ var registry = NewRegistry(
...
+	ContextOptionConfigs []*ContextOptionConfig `json:"contextOptions,omitempty" yaml:"contextOptions,omitempty"`
+	// IncludeFilters controls which resources the metric applies to
+	IncludeFilters []*pb.FlowFilter `json:"includeFilters,omitempty" yaml:"includeFilters,omitempty"`
+	// IncludeFilters controls which resources the metric doesn't apply to
+	ExcludeFilters []*pb.FlowFilter `json:"excludeFilters,omitempty" yaml:"excludeFilters,omitempty"`
+}
+
+// DynamicMetricsConfig represents the structure used for configuring
+// Hubble metrics context options
+type DynamicMetricsConfig struct {
```
Comment (chancez) [inline pkg/hubble/metrics/api/api.go]: Nit: the name here feels a bit off. Whether or not it's dynamic it somewhat orthogonal to the definition itself. Maybe just `Config`?

**Review by chancez — CHANGES_REQUESTED** [review]
Overall LGTM. My only thought is `DynamicMetricsConfig` could probably be renamed since it's really the usage that determines whether or not it's dynamic. It's a bit confusing to have `ParseStaticMetricsConfig` return a `DynamicMetricsConfig`.

File: pkg/hubble/metrics/api/api.go (hunk 1 above)
Comment (rectified95) [reply pkg/hubble/metrics/api/api.go]: Sounds good, renamed to `Config`

Comment (rectified95) [reply pkg/hubble/metrics/api/api.go]: This seems to fail in two ways:
- the linter rejects it with: `api.Config is missing field Metrics` - https://github.com/cilium/cilium/actions/runs/10893022561/job/30227106551?pr=34849
- bpf code fails to build 
`bash: line 1: ccache: command not found make[1]: *** [Makefile:255: bpf_lxc.o] Error 127`

Went with `MetricsConfig`: plural 's' mid-word, for this wrapper type containing a slice of `MetricConfig`s

Comment (chancez) [reply pkg/hubble/metrics/api/api.go]: `exhaustruct` is relatively new, and you probably can indicate the fields are optional to handle that.

The bpf code error is certainly an unrelated issue.

Comment (rectified95) [reply pkg/hubble/metrics/api/api.go]: Added ignore annotations and rebased on main - all resolved.

**Comment by rolinh** [comment]
@rectified95 Could you please squash your commits into one?

**Comment by rectified95** [comment]
Squashed all commits

**Comment by rectified95** [comment]
@chancez let me know if this is good

**Review by chancez — APPROVED** [review]

**Comment by rectified95** [comment]
Can we re-run the 3 failed jobs for extra signal?

Known flake: in Conformance Runtime: https://github.com/cilium/cilium/issues/34731
New issue, similar to other cases of Hubble Relay not starting: https://github.com/cilium/cilium/issues/34994
Maybe existing issue with Hubble Relay connection, this PR in Cilium E2E Upgrade: https://github.com/cilium/cilium/issues/24805#issuecomment-1604381678

**Review by rolinh — APPROVED** [review]
Sorry for the delay in reviewing. Changes lgtm.

**Pull request merged**